digiroad2-viite/src/main/scala/fi/liikennevirasto/viite/process/DefloatMapper.scala

```diff
@@ -0,0 +1,312 @@
+    val startCPAddr = seq.find(_.startCalibrationPoint.nonEmpty)
+    val endCPAddr = seq.find(_.endCalibrationPoint.nonEmpty)
+    if (startCPAddr.nonEmpty) {
+      val (addr, cp) = (startCPAddr.get, startCPAddr.get.startCalibrationPoint.get)
+      if (addr.startAddrMValue != cp.addressMValue)
+        throw new InvalidAddressDataException(s"Start calibration point value mismatch in $cp")
+      if (seq.exists(_.startAddrMValue < cp.addressMValue))
+        throw new InvalidAddressDataException(s"Start calibration point not in the beginning of chain $cp")
+      if (addr.sideCode == SideCode.TowardsDigitizing && Math.abs(cp.segmentMValue) > 0.0 ||
+        addr.sideCode == SideCode.AgainstDigitizing && Math.abs(cp.segmentMValue - addr.endMValue) > 0.001)
+        throw new InvalidAddressDataException(s"Start calibration point LRM mismatch in $cp, sideCode = ${addr.sideCode}, ${addr.startMValue}-${addr.endMValue}")
+    }
+    if (endCPAddr.nonEmpty) {
+      val (addr, cp) = (endCPAddr.get, endCPAddr.get.endCalibrationPoint.get)
+      if (addr.endAddrMValue != cp.addressMValue)
+        throw new InvalidAddressDataException(s"End calibration point value mismatch in $cp")
+      if (seq.exists(_.endAddrMValue > cp.addressMValue))
+        throw new InvalidAddressDataException(s"End calibration point not in the end of chain $cp")
+      if (Math.abs(cp.segmentMValue - addr.endMValue) > 0.1)
+        throw new InvalidAddressDataException(s"End calibration point LRM mismatch in $cp")
+    }
+    val grouped = seq.groupBy(_.linkId).mapValues(_.groupBy(_.sideCode).keySet.size)
+    if (grouped.exists{ case (_, sideCodes) => sideCodes > 1})
+      throw new InvalidAddressDataException(s"Multiple sidecodes generated for links ${grouped.filter(_._2 > 1).keySet.mkString(", ")}")
+  }
+
+  def preTransferChecks(seq: Seq[RoadAddress]): Unit = {
+    if (seq.count(_.startCalibrationPoint.nonEmpty) > 1)
+      throw new IllegalArgumentException("Too many starting calibration points before transfer")
+    if (seq.count(_.endCalibrationPoint.nonEmpty) > 1)
+      throw new IllegalArgumentException("Too many starting calibration points before transfer")
+    val startCPAddr = seq.find(_.startCalibrationPoint.nonEmpty)
+    val endCPAddr = seq.find(_.endCalibrationPoint.nonEmpty)
+    if (startCPAddr.nonEmpty) {
+      val (addr, cp) = (startCPAddr.get, startCPAddr.get.startCalibrationPoint.get)
+      if (addr.startAddrMValue != cp.addressMValue)
+        throw new IllegalArgumentException(s"Start calibration point value mismatch in $cp")
+      if (seq.exists(_.startAddrMValue < cp.addressMValue))
+        throw new IllegalArgumentException("Start calibration point not in the first link of source")
+      if (addr.sideCode == SideCode.TowardsDigitizing && Math.abs(cp.segmentMValue) > 0.0 ||
+        addr.sideCode == SideCode.AgainstDigitizing && Math.abs(cp.segmentMValue - addr.endMValue) > 0.001)
```

Comment:
should we use MaxAllowedMValueError instead of 0.001? Still i dunno if we will be use this const again elsewhere